evm/vm/forks/sharding/guess_head_state_manager.py

```diff
@@ -0,0 +1,245 @@
+            lookahead_period = self.get_current_period() + future_periods
+            result |= self.is_collator_in_period(lookahead_period)
+        return result
+
+    def is_late_collator_period(self):
+        current_period = self.get_current_period()
+        current_block_num = self.vmc.web3.eth.blockNumber
+        last_block_num_in_current_period = current_period * self.vmc.config['PERIOD_LENGTH']
+        return (
+            self.is_collator_in_period(current_period) and
+            (current_block_num + NUM_RESERVED_BLOCKS >= last_block_num_in_current_period)
+        )
+
+    # guess_head process related
+
+    async def verify_collation(self, head_collation_hash, collation_hash):
+        coroutine = fetch_and_verify_collation(
+            collation_hash,
+        )
+        task = asyncio.ensure_future(coroutine)
+        collation_validity = await asyncio.wait_for(task, None)
+        self.collation_validity_cache[collation_hash] = collation_validity
+        if not collation_validity:
+            self.chain_validity[head_collation_hash] = False
+
+    def fetch_candidate_head_hash(self):
+        head_collation_hash = None
+        try:
+            head_collation_dict = self.shard_tracker.fetch_candidate_head()
+            head_collation_hash = head_collation_dict['header'].hash
+        except NoCandidateHead:
+            self.logger.debug("No candidate head available, `guess_head` stops")
+        self.last_period_fetching_candidate_head = self.get_current_period()
+        return head_collation_hash
+
+    def set_collation_task(self, head_collation_hash, current_collation_hash, task):
+        self.unfinished_verifying_tasks.append(task)
+        self.collation_verifying_task[current_collation_hash] = task
+        self.chain_collations[head_collation_hash].append(current_collation_hash)
+
+    def get_chain_tasks(self, head_collation_hash):
+        # ignore tasks that are done
+        return [
+            self.collation_verifying_task[collation_hash]
+            for collation_hash in self.chain_collations[head_collation_hash]
+        ]
+
+    async def guess_head(self, enable_timing_up=False):
+        '''
+        Perform windback process.
+        returns None if there is no candidate head available in this period
+        '''
+        start = time.time()
+
+        head_collation_hash = current_collation_hash = None
+        self.unfinished_verifying_tasks = []
+        while True:
+            # discard old logs if we're in the new period
+            if self.get_current_period() > self.last_period_fetching_candidate_head:
+                self.shard_tracker.clean_logs()
+            head_collation_hash = self.fetch_candidate_head_hash()
+            if head_collation_hash is None:
+                break
+            current_collation_hash = head_collation_hash
+            current_depth = 0
+            while self.chain_validity[head_collation_hash]:
```

Comment:
Feeling a little nervous to see three `while` loops in this function.  And the code snippet:
```python
while self.chain_validity[head_collation_hash]:
    # if running out of time
    if enable_timing_up and self.is_late_collator_period():
        return head_collation_hash
```
is duplicate.

Reply:
Yes you're right. Thank you for pointing out.

Actually I have 2 alternatives for now.
1. Use `for i in range(self.vmc.config['WINDBACK_LENGTH']):` and move the most inner while to the same level as `for`
2. Use two layer `while`, however, in the inner while loop, we don't do `current_collation_hash = self.vmc.get_collation_parent_hash(...)` every time. Instead, need to check if the `current_collation_hash == GENESIS_COLLATION_HASH`, if so, check if the collations of the chain `head_collation` is totally finished.

I prefer the first one since the logic is more clear and readable, I think. And I have a version of this  already.

Do you have any suggestion about which one to choose? or other solutions? Thank you.

Reply:
The code snippet becomes https://gist.github.com/mhchia/8da88064a3efd1014c7d334c8045e370